core/lib/instrument_api.h

```diff
@@ -1806,6 +1806,32 @@ DR_API
 const char *
 dr_get_application_name(void);
 
+DR_API
+/**
+ * Provides information about the app's arguments by setting \p args_buf up to
+ * the size \p buf_size. Returns the number of args set or -1 on error.
+ *
+ * Use dr_app_arg_as_utf8() to get the argument as a string.
+ *
+ * \note Currently, this function is only available on Unix with
+ * early injection.
+ */
+int
+dr_get_app_args(OUT dr_app_arg_t *args_buf, int buf_size);
+
+DR_API
+/* Returns the passed argument as a string. \p buf is used only if needed, and
+ * therefore the caller should not assume that the string is in the \p buf. In other
+ * words, always use the returned value to refer to the string. Returns NULL on error
+ * such as when \p buf is needed as storage and the size of the buffer \p buf_size
+ * is not sufficient.
```

Comment:
Is there a way to tell the caller how much space is needed here as well, or whether to blindly double or give up.

Reply:
You can obtain the original size from dr_app_arg_t. This size should be sufficient no?

Reply:
For Windows, that's the UTF-16 size.  The UTF-8 size is hard to predict from there: you can upper-bound it by 1.5x I guess, but if it's all ascii chars you'll have a 3x-too-big buffer.  Maybe not a big deal.  Of course that means the impl here doesn't know the size unless it does an extra pass through: which io.c supports (utf16_to_utf8_size()).  We could call that extra pass only if we run out of space to avoid wasted work.

Reply:
> We could call that extra pass only if we run out of space to avoid wasted work.

Do mean the API function does this pass or is the pass the responsibility of the caller?

Reply:
The API function.  But if typical usage is to re-use a single buffer for every arg maybe an upper bound is enough?  Maybe just document that the user should allocate a buffer 1.5x the largest dr_app_arg_t.size.  And also have a separate return value for buffer too small?  We never did establish separate error value conventions in the core like most of the extensions have.  Another thing to do differently if we redesigned it all.

Reply:
Doc sounds good. So we can add an OUT bool *flag or rely on the generic return of NULL, which can be returned due to other errors as well. Assuming that the user does nothing weird like passing a NULL buf ptr, then a NULL will typically represent lack of sufficient size. Not really an elegant solution, I know, but which do you prefer?

Reply:
Hmm, and I guess the 3rd choice is to return an error code with the buffer return as an output parameter instead (drmgr/drreg-style), but since no other DR function does that maybe that seems weird.  With the 2nd (output param error code; your "bool *flag" but I'd make it in int error code) or 3rd we'd add a set of error codes: maybe start a trend for new DR API routines to have error codes.  4th choice would be a separate errno()-style function: which we also do for some other interfaces like drcachesim analyzers.

Since we're retrofitting to the existing DR API: I would lean toward choice 4, since we could use that for existing routines too if we want to w/o affecting compatibility.

So sthg like:

`int dr_get_error_code()` which returns the error code of the last DR API routine that failed for the current thread (avoid races: make it per-thread).

And a `DR_ERROR_*` enum?

WDTY?  It may be confusing at first to have this and have only one function using it...but we have to start somewhere.

Reply:
If we add `dr_get_error_code` I would make it *not* be written on success: only on failure?

Reply:
This does sound like a good approach. Performance is just my first quick concern - due to the access of thread local storage.

Reply:
> If we add `dr_get_error_code` I would make it _not_ be written on success: only on failure?

How do you avoid confusion from a previous failure if you do not update?

Reply:
The user has to first check for failure in the API routine they just called, using its normal fail/success return value.

Reply:
Right, so exactly like errno. Cool, makes sense to me.

Reply:
Where is this error code going to be stored please? Recall, we don't have access to dormer's tls array.

Reply:
Was AFK.  A new field inside dcontext->client_data.

Reply:
Sorry for the delay. I had a longer and busier week than usual. 

I still need to add the error coding handling. This is the only thing pending. I'll try to add this soon.

Reply:
Sure, no rush.

Reply:
Done.